### PR TITLE
Improve ANCM file watcher thread management

### PR DESCRIPTION
**Pull request by adityamandaleeka**
I'm trying to get rid of `TerminateThread` usage in ANCM and decided to tackle the file watcher thread first. This change includes various improvements to the file watcher thread management and shutdown mechanisms.

- Creation of a new shutdown event `m_pShutdownEvent` that provides a more explicit and straightforward way to signal the thread to exit.
- Refactoring of `FILE_WATCHER::ChangeNotificationThread` to use `WaitForMultipleObjects` to wait on either the completion port or the new shutdown event, removing the need for the `FILE_WATCHER_SHUTDOWN_KEY` mechanism.
- Simplification of `FILE_WATCHER::WaitForMonitor` by removing calls to `GetExitCodeThread` and `TerminateThread`, relying on `WaitForSingleObject` to handle thread state checking and waiting.
- Replacement of the volatile BOOL `m_fThreadExit` with a `std::atomic<bool>`
- Other minor improvements

Contributes towards https://github.com/dotnet/aspnetcore/issues/45066

### CODE REVIEW

File: src/Servers/IIS/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.cpp [hunk 1]
```diff
@@ -30,33 +36,25 @@ FILE_WATCHER::~FILE_WATCHER()
...
 {
-    if (m_hChangeNotificationThread != NULL)
+    if (m_hChangeNotificationThread == nullptr)
     {
-        DWORD dwExitCode = STILL_ACTIVE;
+        return;
+    }
 
-        while (!m_fThreadExit && dwRetryCounter > 0)
+    while (!m_fThreadExit && dwRetryCounter-- > 0)
```
Comment (BrennanConroy) [inline src/Servers/IIS/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.cpp:44]: What happens if the thread doesn't exit after the loop completes?

And, should we log in that case?

Comment (adityamandaleeka) [reply src/Servers/IIS/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.cpp:44]: Good idea, done.

Comment (adityamandaleeka) [reply src/Servers/IIS/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.cpp:44]: One option if we want to de-risk this in .NET 8 is to enable the termination behavior behind a flag.

Comment (BrennanConroy) [reply src/Servers/IIS/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.cpp:44]: What is the thread doing that could be "bad" if it's still running?

Comment (BrennanConroy) [reply src/Servers/IIS/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.cpp:55]: Offline discussion: The thread is holding onto state that can have undefined behavior if accessed when the class is destructed. e.g. `FILE_WATCHER*` and the `m_pShutdownEvent` and `m_hCompletionPort` events.

Which brings up another thing, those events are never closed via `CloseHandle` it looks like.

File: src/Servers/IIS/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.cpp [hunk 2]
```diff
@@ -9,56 +9,79 @@
...
         if (!m_fThreadExit)
         {
+            LOG_INFO(L"File watcher thread did not exit. Forcing termination.");
             TerminateThread(m_hChangeNotificationThread, 1);
         }
     }
+    else
+    {
+        // Wait for the thread to exit.
+        WaitForSingleObject(m_hChangeNotificationThread, INFINITE);
```
Comment (BrennanConroy) [inline src/Servers/IIS/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.cpp:83]:
```suggestion
        LOG_INFO(L"Waiting for File watcher thread to exit.");
        WaitForSingleObject(m_hChangeNotificationThread, INFINITE);
```
So we know if it's hanging. The thread itself logs when it exits.

**Review by BrennanConroy — APPROVED** [review]
I think a log should be added

**Pull request merged**